### ISSUE TITLE
Invalid disk images

### ISSUE THREAD

**Issue by mlund** [simmons/cbm#1]
Hi and thanks for a great library! I use it for handling disk images for a tool that communicates with the mega65 retro-computer. Many (if not most) disk images from e.g. the mega65 filehost however appear invalid and I wonder if you know why? Here's an example:

- https://files.mega65.org/files/z/Zappy_CdAhgr.d81

**Comment by simmons** [comment]
I've seen some videos of the MEGA65. It looks like a cool project. I'm impressed someone is actually using my old `cbm` crate, and trying to use it with the MEGA65, no less!

When I try to read the directory of the D81 image you supplied, I get the following error:
```
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
```

It looks like there are at least two issues here:
1. The D81 format is considered unformatted by the `cbm` crate. I'm not sure if the MEGA65 is actually trying to use the old CBM 1581 high-level format, or if it uses its own filesystem, or something in between (e.g. some variation on the 1581 format that this crate can't cope with).
2. There is an infinite recursion bug in the `std::fmt::Display` implementation of the `DiskError` enum, so the `DiskError::Unformatted` error is never properly reported.

The infinite recursion bug is trivial to fix. Providing compatibility with MEGA65 disk images will require a bit more research. It will be fun to take a look at this when I get a chance.

**Comment by mlund** [comment]
Thanks for a quick response and that the cbm crate is still alive :-) I think clippy actually warned about the recursion during the 2021 migration, but I didn't pay enough attention.

As far as I know, MEGA65 uses a generic d81 format. I just tried the above d81 image with `c1541` from VICE which does seem to read OK:
~~~
$ c1541 -verbose -attach Zappy_CdAhgr.d81 -validate -dir -info
validating in unit 8 ...
0 "zappy!          " 2022 
27    "zappy!"            prg 
3133 blocks free.
disk format  : 1581
track count  : 80
error block  : No
write protect: Off
~~~

**Comment by simmons** [comment]
It looks like the recursion was part of the 2021 edition PR where it migrated use of the `Error` `description()` method to `to_string()`, and I missed that nuance in the review. At least it wasn't in the crate as published.

The problem with the MEGA65 image is that its disk header doesn't have the expected dos/disk version for 1581 disks (`3D`) at header offsets 0x19 and 0x1A, and the crate is being conservative about parsing images that could possibly be in a different format. From the provided image, the dos and disk type would be `0x32, 0xa0`:
```
00061800: 2803 4400 5a41 5050 5921 a0a0 a0a0 a0a0  (.D.ZAPPY!......
00061810: a0a0 a0a0 a0a0 3230 3232 a0a0 a000 0000  ......2022......
```

The 1581 "Disk DOS version type" at offset 0x02 is correct, though (0x44). I'll need to do some research to see if we should ignore the version bytes, or handle them in a different way. It might be interesting to look at the VICE `c1541` source to see if it's applying special logic for such disks, or if it is taking a "best effort only"/permissive approach. It would also be interesting to look at the MEGA65 DOS ROM to see if there's any insight into how it is constructing the disk header. (I'm guessing that the Zappy image was initialized on a MEGA65.)

**Comment by mlund** [comment]
Thanks for narrowing this down. I'm not sure how the ROM formats, but I have [asked for help in the MEGA65 Discord forum](https://discord.com/channels/719326990221574164/795760112425369610/1046447520533975060).
I'm also unsure of the Zappy disk was formatted on the mega, but I will do some formatting on the machine. Here are a few other images with the issue:
- [megano](https://files.mega65.org?id=9d1e7e80-e0b1-40ad-a6c2-722451d6afc8)
- [char eddie](https://files.mega65.org?id=39993b5a-2816-437f-981c-a02aac8df49e)
- [Bomber Blitz](https://files.mega65.org?id=9a6c653c-93de-41c0-b890-0ce56c39fde2)

**Comment by mlund** [comment]
Creating a new disk image in the "Freezer" menu on the MEGA65 reproduces the issue:
~~~
00061800  28 03 00 00 43 42 4d 54  45 53 54 a0 a0 a0 a0 a0  |(...CBMTEST<A0><A0><A0><A0><A0>|
00061810  a0 a0 a0 a0 00 00 35 37  00 31 44 00 00 00 00 00  |<A0><A0><A0><A0>..57.1D.....|
00061820  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
~~~
Using `hexdump -C -s 0x61819 -n 2`, so far I found the following bytes at header offset 0x19-0x1a:
- 0x33 0x44 ("3D" and thus OK)
- 0x32 0xa0 ("2?")
- 0x32 0x32 ("22")
- 0x31 0x44 ("1D" generated by MEGA65 freezer)

**Comment by mlund** [comment]
Here's the [reply from Discord](https://discord.com/channels/719326990221574164/795760112425369610/1046463473225322516):
_c1541 and cc1541 let you set an ID of 5 characters (instead of the normal 2) that starts at 0x16 (so the last 2 characters will be overwriting 0x19 and 0x1a).
The default is '00 2a' or '00 3d', but there's nothing stopping you from putting the current year (' 2022') in there, in which case the value at 0x19 and 0x1a will be 32 32._

**Comment by simmons** [comment]
Making the cbm crate less restrictive about the format type field (or "dos/disk version" or whatever -- the name of this field seems to vary from place to place) is easy. But figuring out the right thing to do -- whether to make it more permissive, and how much more permissive to make it -- is a tougher job.

I've been thinking about this, and I'm now ~85% confident that we can safely make the crate more permissive. (I can't be any more confident without performing a full reverse engineering of all CBM ROMs and tools that may work with disk images.) I will look into making this change this weekend.

There are two places where CBM DOS ROMs will store metadata about the DOS version/format:
1. **Disk DOS version type (offset 0x02 in D81 header).** This should always be `0x44` in D81 images. The CBM DOS will actively verify this value during disk operations, and change its behavior if it is anything different. (If the value is different, it will refuse to write to the disk. Some people have used this as a primitive form of write protection.) We **should** continue to verify this value in disk images, as any different value likely indicates that the disk image has not been high-level formatted.
2. **Format type (offsets 0x19-0x1A in D81 header).** This is usually `0x33 0x44` (`"3D"`) on disks formatted by a 1581 drive. I've glanced at a [1581 DOS ROM disassembly](http://www.zimmers.net/anonftp/pub/cbm/src/drives/1581-rom-source.zip), and this value seems to only ever be written to the disk, and never actually consumed by the ROM. Therefore, this field is merely cosmetic -- it only exists for the benefit of a human operator who would like some insight into what DOS was used to format the disk. A C64 with a 1581 drive will happily read any disk regardless of what value is present here. (I've verified that a C64/1581 has no problems reading the directory and file from your sample Zappy D81 image.) Therefore, we **should not** continue to verify this value in disk images. Now that we are dealing with a wider range of DOS ROMs (e.g. the C65/MEGA65), we should expect to encounter divergent values here.

In dealing with disk images associated with C65/MEGA65 systems, we see two notable divergences from the format type as written by the 1581 drive:
1. **Format type 1D**. From my emulator tests with C65 ROM 910828 and MEGA65 ROM 990003, these systems always write `0x31 0x44` (`"1D"`) into the format type field. The `cbm` crate currently rejects such disk images. I'm not sure why the Commodore engineers seemingly chose to use a lower value for this newer system, but maybe it's a reference to the last two characters of the C65's `F011D` disk controller chip.
2. **Arbitrary format type.** The Zappy D81 disk image you provided completely ignores the CBMFS header format and writes `"2002   "` into the bytes normally occupied by the disk ID, the format type, and the surrounding padding. The `cbm` crate currently rejects such disk images. As you mentioned from the Discord conversation, this is presumably a result of the c1541/cc1541 tools playing fast and loose with the format. This is **not** a valid use of the CBMFS format *. However, it just happens to work since the CBM DOS ROMs ignore this field, and the rendered directory listing (as tokenized BASIC when loading `$`) naively includes all the bytes. Since it's a lucky coincidence that it works, I'm not surprised people would try to be creative with these bytes, since people have always tried to push the limits of these 8-bit systems.

\* Note: It may be somewhat a matter of opinion about what is and isn't standard CBMFS, since these standards were never specified with great rigor. Even the [*Inside Commodore DOS* book](https://www.lyonlabs.org/commodore/onrequest/Inside_Commodore_Dos.pdf) from 1984 seems reluctant to go into great detail about these things.

For reference, I've collected some D81 disk header dumps from various disks:

A C64 / 1581 freshly formatted disk with name `MYDISKNAME` and ID `01`.
```
00061800: 2803 4400 4d59 4449 534b 4e41 4d45 a0a0  (.D.MYDISKNAME..
00061810: a0a0 a0a0 a0a0 3031 a033 44a0 a000 0000  ......01.3D.....
00061820: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00061830: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

C65 freshly formatted disk with `HEADER "MYDISKNAME", I01`.
```
00061800: 2803 4400 4d59 4449 534b 4e41 4d45 a0a0  (.D.MYDISKNAME..
00061810: a0a0 a0a0 a0a0 3031 a031 44a0 a000 0000  ......01.1D.....
00061820: 0151 0000 0000 0000 0000 0000 0000 0000  .Q..............
00061830: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

MEGA65 freshly formatted disk with `HEADER "MYDISKNAME"`. In the emulator, I was unable to successfully include the ID to ask for a low-level format, and it somehow chose the id `BS`.
```
00061800: 2803 4400 4d59 4449 534b 4e41 4d45 a0a0  (.D.MYDISKNAME..
00061810: a0a0 a0a0 a0a0 4253 a031 44a0 a000 0000  ......BS.1D.....
00061820: 0151 0000 0000 0000 0000 0000 0000 0000  .Q..............
00061830: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

The problematic Zappy disk image, which completely disregards the CBMFS header format and writes the current year (2022) into the bytes that would otherwise by used by the ID and format type.
```
00061800: 2803 4400 5a41 5050 5921 a0a0 a0a0 a0a0  (.D.ZAPPY!......
00061810: a0a0 a0a0 a0a0 3230 3232 a0a0 a000 0000  ......2022......
00061820: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00061830: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

The C65 development system included several sample disks. Curiously, these are formatted with a format type of `3D`, as if they were written by a 1581 drive:
```
00061800: 2803 4400 4336 3520 4449 534b 2031 a0a0  (.D.C65 DISK 1..
00061810: a0a0 a0a0 a0a0 3531 a033 44a0 a000 0000  ......51.3D.....
00061820: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00061830: 0000 0000 0000 0000 0000 0000 0000 0000  ................
```

**Comment by mlund** [comment]
Thanks for this thorough investigation - this is very useful also for other tools handling cbm disks. Perhaps the information could eventually be included in the documentation? I have access to the latest MEGA65 ROM and can do some testing too. Also, I should be able to access the source code, but will need to fix that. As you mention, the "format" is often violated with all sorts of tricks, but as long as the ROM can handle it, I personally think it's OK as this was very common. Perhaps the crate could provide information about the disk type, i.e. "1D", "3D", or "unknown" to allow for an optional more lax interpretation.

**Comment by mlund** [comment]
Here's output from the mega65 rom 920377 which is the current stable version. The image is made with Xemu which generates a random ID, here `4M`, that does not change with the `HEADER` command. Unfortunately I don't have access to mega65 hardware for the next month (should have stuffed it in the backpack). `BS` in the earlier rom likely refers to [_Bit Shifter_](https://github.com/Edilbert) which is the main MEGA65 ROM developer.


- `HEADER "MYDISKNAME"`:
  ~~~
  00061800  28 03 44 00 52 4f 4d 39  32 30 33 37 37 a0 a0 a0  |(.D.ROM920377...|
  00061810  a0 a0 a0 a0 a0 a0 34 4d  a0 31 44 a0 a0 00 00 00  |......4M.1D.....|
  00061820  01 51 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |.Q..............|
  00061830  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
  ~~~

**Comment by simmons** [comment]
> Perhaps the information could eventually be included in the documentation?

Yes, I've done more research and provided a summary of the findings in the `cbm::disk` [module documentation](https://github.com/simmons/cbm/pull/3/files#diff-95d7544cbb578d89eaaed66c11ce9f6b3695b3ef56508092ba56c39b1a6cb90cR18) in PR #3.

> I have access to the latest MEGA65 ROM and can do some testing too. Also, I should be able to access the source code, but will need to fix that. As you mention, the "format" is often violated with all sorts of tricks, but as long as the ROM can handle it, I personally think it's OK as this was very common. 

I agree. Removing the requirement for the field I formerly called the "format type" made most of the above linked MEGA65 sample disk images work. However, the `CEV1_4rXpHp.D81` disk image ("char eddie") still did not work. This image used a diskette DOS version value of `0x00`, which it turns out is valid for both reading and writing in the CBM DOS ROMs I've studied. Therefore, I need to revise my earlier comment about this field:

> **Disk DOS version type (offset 0x02 in D81 header)**. ...  We **should** continue to verify this value in disk images, as any different value likely indicates that the disk image has not been high-level formatted.

We actually **should not** try to verify this field.

I've done more research on this topic, and have performed some software archaeology on the 1540/1541, 1571, 1581, and C65 DOS ROMs. I think I now have a much clearer picture of the nature of these fields and the intentions of the original Commodore engineers.

Sadly, I think I've been using the wrong terminology for these fields. I've been mistakenly treating the directory DOS version and directory format type as a single combined field. (For example, referring to the C65 "format type" as "1D".) I was biased from my years as a _user_ of these systems, where these fields were always presented and processed as a single, combined two-character string. Indeed, even the sample program from the 1541 user's guide refers to these two characters simply as the "operating system string" [1], and indeed reads them as a single string field. I think we should instead use the terminology found in the comments of the 1540 CBM DOS ROM source by the original Commodore engineers -- "diskette format type", "directory DOS version", and "directory format type" -- since it makes a clear distinction between the format type as written to offset 0x02 (“diskette format type”) and the format type as written to offset 0xA6 (1541/1571) or 0x1A (1581) (“directory format type”).

I've opened PR #3 which removes expectation enforcement for all three of these fields, adds documentation, and revises the field names.
